doc/users-guide/datapar/promotion.rst

```diff
@@ -0,0 +1,324 @@
+  :language: chapel
+  :lines: 105-106
+  :dedent: 2
+
+Attempting to compile this promoted call results in a compile-time
+error.  To see the reason, consider the equivalent forall-loop:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 108-109
+  :dedent: 2
+
+On the way into the loop, a ``const`` shadow copy of *r* is created as
+with any other forall-loop.  But since constants can't be passed by
+``ref`` argument intent, an error is triggered.  As with forall
+intents on explicit loops, these rules prevent common race conditions.
+For example, if the call had been executed, it would be a race as to
+which of *B*'s values would be stored into *r*.
+
+Promoted calls do not support a mechanism to override default forall
+intents.  Thus, in order to override the default and enable the race,
+a user would need to write the loop out explicitly and add a ``with``
+clause as follows:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 114-115
+
+
+Promoting Using Ranges and Domains
+----------------------------------
+
+As mentioned at the outset, not only can arrays be used to promote
+functions, but ranges and domains can as well.  For example, the
+following two calls promote ``maybeCopy()`` using a range and a
+domain, respectively:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 185,189
+
+The contents of *A* after these calls is as follows:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.good
+  :language: text
+  :lines: 45,46
+
+At the time of this writing, Chapel does not support an official way
+for users to create their own collection types that support promotion.
+We expect to support this in the future by having the collection type
+support certain well-defined iterator methods.
+
+
+Promoted Library Routines and Operators
+---------------------------------------
+
+All of the examples above illustrate promotion on procedures written
+by a user.  Yet any procedure is a candidate for promotion in Chapel,
+including library routines and operators.  As an example, consider the
+following lines of code:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.chpl
+  :language: chapel
+  :lines: 119-120,127
+
+The second statement promotes the standard ``sqrt()`` library routine,
+resulting in an array of results, stored in *B*:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.good
+  :language: text
+  :lines: 32
+
+The third statement doubles the array *A* and stores the result in *B*
+as expected:
+
+.. literalinclude:: examples/users-guide/datapar/promotion.good
+  :language: text
+  :lines: 34
+
+Note that Chapel does not support an explicit ``+`` operator for
+arrays of reals.  Rather, Chapel implements this operation by
+promoting the standard ``+`` operator for scalar values of type
+``real``.  For this reason, applying ``*`` to Chapel arrays results in
```

Comment:
I was initially confused by the switch from ``+`` to ``*``. Maybe instead of "For this reason" it should say "Similarly, ``*`` on arrays is implemented with promotion. For this reason, ..."